### PR TITLE
Give better messages when asserts in InstallPlan fail.

### PR DESCRIPTION
**Pull request by ezyang**
Ideas:

- Pass Maybe String instead of Bool to record information
  about failure.

- Use assertMsg to do an assert failure with message.
  We have the same semantics as assert by using the
  'assert' combinator, but raising an async exception
  when it forces the bool (and it would be false.)

Signed-off-by: Edward Z. Yang <ezyang@cs.stanford.edu>

### CODE REVIEW

File: cabal-install/Distribution/Client/InstallPlan.hs [hunk 1]
```diff
@@ -562,12 +565,12 @@ failed plan (Processing processingSet completedSet failedSet) pkgid =
...
     assert (all (`Set.notMember` completedSet)  (tail newlyFailedIds)) $
     assert (all (`Set.notMember` failedSet)     (tail newlyFailedIds)) $
-    assert (processingInvariant plan processing') $
+    assertMaybeMsg (processingInvariant plan processing') $
 
     ( map asConfiguredPackage (tail newlyFailed)
     , processing' )
   where
-    processingSet' = Set.delete pkgid processingSet
+    processingSet' = processingSet `Set.difference` Set.fromList newlyFailedIds
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/InstallPlan.hs]: Was this meant to be a part of the patch?

**Comment by 23Skidoo** [comment]
LGTM modulo one small comment.

**Comment by 23Skidoo** [comment]
Also fails on 7.10.

File: cabal-install/Distribution/Client/InstallPlan.hs (hunk 1 above)
Comment (dcoutts) [inline cabal-install/Distribution/Client/InstallPlan.hs]: Ah, I suspect that is a fix for a bug that was triggering the assertion failure...

Now that I see it, it's clear, we don't just remove the single failed package from the processing set but all the failed ones (the failure + the things that depend on the one that failed).

But yes, I expect @ezyang meant to put that fix in a separate patch.

**Comment by dcoutts** [comment]
Yep, apart from putting that fix into a separate patch, all LGTM.

**Comment by dcoutts** [comment]
So we discovered (#3889) that the "fix" is wrong, and it's the invariant that's wrong. I'll push a fix to the invariant, plus a couple other invariant extensions we do think are true.

**Comment by dcoutts** [comment]
@ezyang unless you much prefer this approach to the one I propose in #3901 then I suggest we close this.

**Comment by ezyang** [comment]
Well, I personally think the "assertMsg" combinator is a useful one; I'd rather have my asserts by default output enough information for me to debug them, than having to edit cabal-install to add enough tracing.  It's even more important if a user gets an assert failure and doesn't know how to repro. But yes I'll close this for now.

**Comment by dcoutts** [comment]
> It's even more important if a user gets an assert failure and doesn't know how to repro.

The assert failures have line numbers so it should be enough for a developer to figure out.

**Comment by ezyang** [comment]
I mean, *sometimes* I look at the assert that failed and think, "Aha! It's obvious." But not all the time. I appreciate the extra output in those cases.

**Pull request closed without merging**